File: android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSMessageAction.kt
package live.hms.hmssdk_flutter
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import live.hms.video.sdk.*
import live.hms.video.error.HMSException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import live.hms.hmssdk_flutter.HMSErrorLogger.Companion.returnError
import live.hms.video.sdk.models.HMSMessage

class HMSMessageAction {
    companion object {
        fun messageActions(call: MethodCall, result: Result,hmssdk:HMSSDK) {
            when (call.method) {
                "send_broadcast_message" -> {
                    sendBroadCastMessage(call, result,hmssdk)
                }
                "send_direct_message" -> {
                    sendDirectMessage(call, result,hmssdk)
                }
                "send_group_message" -> {
                    sendGroupMessage(call, result,hmssdk)
                }
                else -> {
                    result.notImplemented()
                }
            }
        }


        private fun sendBroadCastMessage(call: MethodCall, result: Result,hmssdk:HMSSDK) {
            val message = call.argument<String>("message") ?:
            run{
                HMSErrorLogger.logError("sendBroadCastMessage", "message is null", "Parameter Error")
                result.success(HMSExceptionExtension.getError("message is null in sendBroadCastMessage"))
                return
            }
            val type = call.argument<String>("type") ?: "chat"
            hmssdk.sendBroadcastMessage(message, type, getMessageResultListener(result))
        }

        private fun sendGroupMessage(call: MethodCall, result: Result,hmssdk:HMSSDK) {
            val message = call.argument<String>("message") ?:
            run{
                HMSErrorLogger.logError("sendGroupMessage", "message is null", "Parameter Error")
                result.success(HMSExceptionExtension.getError("message is null in sendGroupMessage"))
                return
            }
            val roles: List<String>? = call.argument<List<String>>("roles")
            val type = call.argument<String>("type") ?: "chat"

            val hmsRoles = hmssdk.getRoles().filter { roles?.contains(it.name)!! }

            hmssdk.sendGroupMessage(message, type, hmsRoles, getMessageResultListener(result))
        }

        private fun sendDirectMessage(call: MethodCall, result: Result,hmssdk:HMSSDK) {
            val message = call.argument<String>("message") ?:
            run{
                HMSErrorLogger.logError("sendDirectMessage", "message is null", "Parameter Error")
                result.success(HMSExceptionExtension.getError("message is null in sendDirectMessage"))
                return
            }
            val peerId = call.argument<String>("peer_id") ?:
            run {
                HMSErrorLogger.logError("sendDirectMessage", "peerId is null", "Parameter Error")
                result.success(HMSExceptionExtension.getError("peerId is null in sendDirectMessage"))
                return
            }
            val type = call.argument<String>("type") ?: "chat"
            val peer = HMSCommonAction.getPeerById(peerId,hmssdk)
            peer ?: run{
                HMSErrorLogger.logError("sendDirectMessage", "peer is null", "changeRole Error")
                result.success(HMSExceptionExtension.getError("peer is null in sendDirectMessage"))
                return
            }
            hmssdk.sendDirectMessage(message, type, peer, getMessageResultListener(result))
        }

        private fun getMessageResultListener(result: Result) = object: HMSMessageResultListener {
            override fun onError(error: HMSException) {
                val args = HashMap<String, Any?>()
                args["event_name"] = "on_error"
                args["data"] = HMSExceptionExtension.toDictionary(error)
                if (args["data"] != null)
                    CoroutineScope(Dispatchers.Main).launch {
                        result.success(args)
                    }
            }

            override fun onSuccess(hmsMessage: HMSMessage) {
                val args = HashMap<String, Any?>()
                args["event_name"] = "on_success"
                args["message"] = HMSMessageExtension.toDictionary(hmsMessage)
                if (args["message"] != null)
                    CoroutineScope(Dispatchers.Main).launch {
                        result.success(args)
                    }
            }
        }

    }
}

File: ios/Classes/Actions/HMSMessageAction.swift
//
//  HMSMessageAction.swift
//  hmssdk_flutter
//
//  Created by govind on 03/02/22.
//

import Foundation
import HMSSDK

class HMSMessageAction {
    static func messageActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        switch call.method {
        case "send_broadcast_message":
            sendBroadcastMessage(call, result, hmsSDK)

        case "send_direct_message":
            sendDirectMessage(call, result, hmsSDK)

        case "send_group_message":
            sendGroupMessage(call, result, hmsSDK)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    static private func sendBroadcastMessage(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        let arguments = call.arguments as! [AnyHashable: Any]

        guard let message = arguments["message"] as? String
        else {
            result(HMSErrorExtension.getError("No message found in \(#function)"))
            return
        }

        let type = arguments["type"] as? String ?? "chat"

        hmsSDK?.sendBroadcastMessage(type: type, message: message) {data, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                guard let data
                else {
                    result(nil)
                    return
                }
                var dict = [String: Any]()
                dict["message"] = HMSMessageExtension.toDictionary(data)
                result(dict)
            }
        }
    }

    static private func sendDirectMessage(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        let arguments = call.arguments as! [AnyHashable: Any]

        guard let message = arguments["message"] as? String,
              let peerID = arguments["peer_id"] as? String,
              let peer = HMSCommonAction.getPeer(by: peerID, hmsSDK: hmsSDK)
        else {
            result(HMSErrorExtension.getError("Invalid arguments passed in \(#function)"))
            return
        }

        let type = arguments["type"] as? String ?? "chat"

        hmsSDK?.sendDirectMessage(type: type, message: message, peer: peer) { data, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                guard let data
                else {
                    result(nil)
                    return
                }
                var dict = [String: Any]()
                dict["message"] = HMSMessageExtension.toDictionary(data)
                result(dict)
            }
        }
    }

    static private func sendGroupMessage(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        let arguments = call.arguments as! [AnyHashable: Any]

        guard let message = arguments["message"] as? String,
              let rolesList = arguments["roles"] as? [String],
              let roles: [HMSRole] = (hmsSDK?.roles.filter { rolesList.contains($0.name) })
        else {
            result(HMSErrorExtension.getError("Invalid arguments passed in \(#function)"))
            return
        }

        let type = arguments["type"] as? String ?? "chat"

        hmsSDK?.sendGroupMessage(type: type, message: message, roles: roles) { data, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                guard let data
                else {
                    result(nil)
                    return
                }
                var dict = [String: Any]()
                dict["message"] = HMSMessageExtension.toDictionary(data)
                result(dict)
            }
        }
    }
}
